components/CharmCalculator.tsx

```diff
@@ -0,0 +1,198 @@
+
+import { Charm, CharmPosition, StatSummary, GetStatSummary } from "@/lib/charm";
+
+export default function CharmCalculator() {
+  const availableClasses = Heroes.Classes.map((classType) => classType.Class);
+  const [selectedClass, setSelectedClass] = useState<String>(availableClasses[0]);
+
+  const [availableHeroes, setAvailableHeroes] = useState<any[] | undefined>(undefined);
+  const [selectedHero, setSelectedHero] = useState<String | undefined>(undefined);
+
+  const [availableCharms, setAvailableCharms] = useState<Charm[] | undefined>(undefined);
+  const [selectedLeftCharm, setSelectedLeftCharm] = useState<Charm | undefined>(undefined);
+  const [selectedMidCharm, setSelectedMidCharm] = useState<Charm | undefined>(undefined);
+  const [selectedRightCharm, setSelectedRightCharm] = useState<Charm | undefined>(undefined);
+
+  const [statsSummary, setStatsSummary] = useState<StatSummary[]>([]);
+
+  /** Sets the current class and clears the currently selected hero and available charms
+   *  Sets available heroes based on selected class
+   *
+   * @param chosenClass | String
+   */
+  function setClass(chosenClass: String): void {
+    // set class, then set available heroes based on that class
+    setSelectedClass(chosenClass);
+    setSelectedHero(undefined);
+
+    Heroes.Classes.find((classType) => {
+      if (classType.Class === chosenClass) {
+        setAvailableHeroes(classType.Heroes.map((hero) => hero.Name));
+      }
+    });
+  }
+
+  /** Sets the current hero and clears the selected charms
+   *  Sets the available charms based on selected hero
+   *
+   * @param heroName | String
+   */
+  function setHero(heroName: String): void {
+    // set hero, then get list of available charms based on that hero
+    setSelectedHero(heroName);
+    setSelectedLeftCharm(undefined);
+    setSelectedMidCharm(undefined);
+    setSelectedRightCharm(undefined);
+
+    Heroes.Classes.find((classType) => {
+      if (classType.Class === selectedClass) {
+        classType.Heroes.find((hero) => {
+          if (hero.Name === heroName) {
+            let charms: Charm[] = [];
+            hero.Charms.forEach((charm) => {
+              charms.push(new Charm(charm.Name, charm.Stats));
+            });
+            setAvailableCharms(charms);
+          }
+        });
+      }
+    });
+  }
+
+  /** Sets the charm that you selected for the given position
+   *  Any time a charm is changed, recalculate the summary
+   *
+   * @param charmName | String
+   * @param position | CharmPosition (enum) | Left/Mid/Right
+   */
+  function setCharm(charmName: string, position: CharmPosition): void {
+    // Get charm and set that selectedCharm to the current charm.
+    availableCharms?.find((charm) => {
+      if (charm.name === charmName) {
+        switch (position) {
+          case CharmPosition.Left:
+            setSelectedLeftCharm(charm);
+            setStatsSummary(GetStatSummary(charm, selectedMidCharm, selectedRightCharm));
```

Comment:
if you change to using a single state variable for all 3 selected charms, then you can utilize `useEffect`:

```js
useEffect(() => {
  setStatsSummary(GetStatsSummary(selectedCharms))
}, [selectedCharms])
```

otherwise you'd do this:
```js
useEffect(() => {
  setStatsSummary(GetStatsSummary(selectedLeftCharm, selectedMidCharm, selectedRightCharm))
}, [selectedLeftCharm, selectedMidCharm, selectedRightCharm])
```